common/ffsbuilder.cpp

```diff
@@ -379,34 +482,966 @@ USTATUS FfsBuilder::buildFreeSpace(const UModelIndex & index, UByteArray & freeS
+                                if (!result) {
+                                    // Calculate volume base
+                                    volumeBase = imagebase - relbase;
+                                    baseFound = true;
+                                    goto out;
+                                }
+                            }
+                            sectionOffset += model->header(peiFile.child(j, 0)).size() + model->body(peiFile.child(j, 0)).size();
+                            sectionOffset = ALIGN4(sectionOffset);
+                        }
+                    }
+                    fileOffset += model->header(index.child(i, 0)).size() + model->body(index.child(i, 0)).size();
+                    fileOffset = ALIGN8(fileOffset);
+                }
+            }
+out:
+            // Do not set volume base
+            if (!baseFound)
+                volumeBase = 0;
+
+            // Reconstruct files in volume
+            UINT32 offset = 0;
+            UByteArray padFileGuid = EFI_FFS_PAD_FILE_GUID;
+            UByteArray vtf;
+            UModelIndex vtfIndex;
+            UINT32 nonUefiDataOffset = 0;
+            UByteArray nonUefiData;
+            for (int i = 0; i < model->rowCount(index); i++) {
+                // Inside a volume can be files, free space or padding with non-UEFI data
+                if (model->type(index.child(i, 0)) == Types::File) { // Next item is a file
+
+                    // Align to 8 byte boundary
+                    UINT32 alignment = offset % 8;
+                    if (alignment) {
+                        alignment = 8 - alignment;
+                        offset += alignment;
+                        volume += UByteArray(alignment, empty);
+                    }
+
+                    // Calculate file base
+                    UINT32 fileBase = volumeBase ? volumeBase + header.size() + offset : 0;
+
+                    // Reconstruct file
+                    result = buildFile(index.child(i, 0), volumeHeader->Revision, polarity, fileBase, file);
+                    if (result)
+                        return result;
+
+                    // Empty file
+                    if (file.isEmpty())
+                        continue;
+
+                    EFI_FFS_FILE_HEADER* fileHeader = (EFI_FFS_FILE_HEADER*)file.data();
+                    UINT32 fileHeaderSize = sizeof(EFI_FFS_FILE_HEADER);
+                    if (volumeHeader->Revision > 1 && (fileHeader->Attributes & FFS_ATTRIB_LARGE_FILE))
+                        fileHeaderSize = sizeof(EFI_FFS_FILE_HEADER2);
+
+                    // Pad file
+                    if (fileHeader->Type == EFI_FV_FILETYPE_PAD) {
+                        padFileGuid = file.left(sizeof(EFI_GUID));
+
+                        // Parse non-empty pad file
+                        if (model->rowCount(index.child(i, 0))) {
+                            //TODO: handle it
```

Comment:
From what I remember, there was an issue with some firmwares expecting data in pad files. In the old version there was a warning regarding pad file contents being destroyed, might it need to be added here in addition to the todo comment, or it is already handled elsewhere?